Improve `-debug-constraints` output for key path components

beccadax opened the pull request:
During constraint debugging, the AST dumper currently shows all `KeyPathExpr` components with `<null>` types because it can't look into the constraint system to find type variables. This PR improves this debugging experience, which should help us fight some of the bugs in key path type-checking.

- [x] Make key path component type variables consistently accessible by constraint locator
- [x] Add callback for ASTDumper to find them
- [x] Colorize, rearrange, and otherwise improve key path component AST dump formatting

**Before**

<img width="1020" alt="Screen Shot 2019-04-16 at 9 22 44 PM" src="https://user-images.githubusercontent.com/931226/56260897-db745400-608d-11e9-9d94-8053e60e0876.png">

**After**

<img width="1020" alt="Screen Shot 2019-04-16 at 9 22 10 PM" src="https://user-images.githubusercontent.com/931226/56260908-e4fdbc00-608d-11e9-9e18-96fdedcb648b.png">

beccadax committed:
[ConstraintSolver] Centralize dumping an expr

There are two places where we can dump an expr with types from the constraint solver. There should only be one.

beccadax committed:
[ConstraintSolver] Dump key path component types

ASTDumper doesn’t have any way to look up key path component types in the constraint solver, so they’re currently shown as null. This change adds a hook to look them up and looks in the key path component’s FunctionResult locator, which is where subscripts already keep their return type.

beccadax committed:
[ConstraintSolver] Give type vars for keypath components useful locators

Previously, nearly all of the type vars for a key path component were dumped under the KeyPathExpr itself, whcih made them difficult to differentiate when debugging. Attach each one to “key path component #N -> function result” instead.

# Conflicts:
#	lib/Sema/CSGen.cpp

beccadax committed:
Reformat and colorize KeyPathExpr AST dumps

brentdax commented:
@swift-ci please smoke test

xedin reviewed (approved):
LGTM! But I'd prefer we didn't iterate over type variables to find one associated with each component, that could be done separately though.

xedin commented on lib/Sema/CSGen.cpp:
```diff
@@ -2930,6 +2930,11 @@ namespace {
+        auto memberLocator = CS.getConstraintLocator(
+            locator, ConstraintLocator::PathElement::getKeyPathComponent(i));
+        auto resultLocator = CS.getConstraintLocator(
+            memberLocator, ConstraintLocator::FunctionResult);
```
Nit: `FunctionResult` is not the best element for the result of the property lookup :)

brentdax replied:
@xedin It's the one already used by subscripts, which is why I chose it. But I don't have to do it that way. What would be an appropriate path element that's not currently used for anything on key path components?

xedin replied:
I think you can create a `ComponentResult` or locator element per component like `KeyPathPropertyResult`

beccadax committed:
[ConstraintSystem] Add KeyPathComponentTypes map

Avoids digging through the long list of type vars to find the ones related to key path components when dumping an expression.

beccadax committed:
[ConstraintSystem] Add a KeyPathComponentResult locator path element

We’ll use this instead of the function result, except for subscript components, which are already structured similarly to subscript expressions.

brentdax commented on lib/AST/ASTDumper.cpp:
```diff
@@ -2723,8 +2731,9 @@ void Expr::dump() const {
 void Expr::dump(raw_ostream &OS,
                  llvm::function_ref<Type(const Expr *)> getTypeOfExpr,
                  llvm::function_ref<Type(const TypeLoc &)> getTypeOfTypeLoc,
+                 llvm::function_ref<Type(const KeyPathExpr *E, unsigned index)> getTypeOfKeyPathComponent,
```
Rather than scattering `function_ref`s like this throughout Expr.h, we should refactor to add some kind of `ExprTypeMap` type with these kinds of operations. The base class would just call through to the AST node, but `ConstraintSystem` could derive from it and override them to look into its internal tables. Then we should rip out all the `getType`, `setType`, etc. closures we pass around and use that instead.

That's a bigger job than I'm prepared to do in this PR, though.

xedin replied:
That's fair!

xedin reviewed (approved):
LGTM, thank you!

brentdax commented:
@swift-ci please smoke test and merge

brentdax commented:
@swift-ci please smoke test and merge

brentdax commented:
Missed a bug, and simplified history while I was at it.

brentdax commented:
@swift-ci please smoke test

brentdax commented:
@swift-ci please smoke test linux platform

xedin commented:
@swift-ci test source compatibility

xedin commented:
@swift-ci please smoke test Linux platform

xedin commented:
@swift-ci please test source compatibility

brentdax commented:
@swift-ci please smoke test Linux platform

brentdax commented:
@swift-ci please smoke test Linux platform
